evm_loader/program/src/account_storage.rs

```diff
@@ -29,75 +29,65 @@ impl<'a> ProgramAccountStorage<'a> {
         debug_print!("account_storage::new");
         let mut accounts = Vec::with_capacity(account_infos.len());
         let mut aliases = Vec::with_capacity(account_infos.len());
+        let mut account_metas = Vec::with_capacity(account_infos.len());
 
         let mut contract_id: H160 = H160::zero();
         let mut caller_id: H160 = H160::zero();
 
-        let mut skip_next: bool = false;
-        let mut contract: bool = true;
-        let mut caller: bool = true;
-        for (i, account_info) in (&account_infos).iter().enumerate() {
-            debug_print!("{}", &i);
-
-            if skip_next {
-                skip_next = false;
-                accounts.push(None);
-                continue;
-            }
+        let account_info_iter = &mut account_infos.iter();
+        let mut index: usize = 0;
 
+        while let Ok(account_info) = next_account_info(account_info_iter) {
             if account_info.owner == program_id {
                 let account_data = AccountData::unpack(&account_info.data.borrow())?;
-                let account = match account_data {
-                    AccountData::Account(ref acc) => acc,
-                    _ => return Err(ProgramError::InvalidAccountData),
-                };
+                let account = AccountData::get_account(&account_data)?;
+
                 let code_data = if account.code_account == Pubkey::new_from_array([0u8; 32]) {
-                    debug_print!("code_account == Pubkey::new_from_array([0u8; 32])");
+                    debug_print!("Common account");
+
+                    if caller_id == H160::zero() {
+                        caller_id = account.ether;
+                        debug_print!("caller id: {}", &caller_id.to_string());
+                    }
```

Comment:
`caller_id` will be the first non-code account. What happens if the first account doesn't belong to program_id? Or with code?

I suggest we need strictly limit the order of caller and contract accounts (as described in `Call` instruction).
Moreover, the contract always should contain the code part and the caller should be **without** code.